### ISSUE TITLE
Example of C++ API usage

### ISSUE THREAD

**Issue by CogMachines** [apple/turicreate#211]
I have trained and deployed a model using turicreate in a shipping iOS app (www.kompapp.com).  Our notation and classification engine is in C++.  With our previous classifier, we could access the core code using Obj-C and C++, allowing us to have both an iOS and server-based solution.  I have been able to classify in C++ using the turi model by embedding Python and using PyObject calls.  This seems overly complex and is also leading to memory corruption issues, as I am probably abusing the Python runtime.

Is there is simple example of loading a trained model using the C++ API and performing a classification?  This would greatly help us.  

Thanks for making your great product available to use!

**Comment by znation** [comment]
Hi @CogMachines, I uploaded the [C++ docs](https://apple.github.io/turicreate/docs/cpp/). Since those can be kind of unapproachable (more of a class/function reference than a howto -- the search works well, but I don't tend to browse), I also wanted to add some sample code here.

Let's assume we're starting with the model that results from the introductory example in the [boosted trees classifier](https://github.com/apple/turicreate/blob/3490286b27ff5d79cb90d09fe026d5671ce990c7/userguide/supervised-learning/boosted_trees_classifier.md#gradient-boosted-regression-trees) section of the user guide. The model is available as [mushroom.tcmodel.zip](https://github.com/apple/turicreate/files/1654991/mushroom.tcmodel.zip) as well.

To make predictions from this model in Python, you could run:
```python
import turicreate as tc
data =  tc.SFrame.read_csv('turicreate/lfs/datasets/xgboost/mushroom.csv')
model = tc.load_model('temp4/mushroom.tcmodel')
data[0]
```
```
{'bruises?': 't',
 'cap-color': 'n',
 'cap-shape': 'x',
 'cap-surface': 's',
 'gill-attachment': 'f',
 'gill-color': 'k',
 'gill-size': 'n',
 'gill-spacing': 'c',
 'habitat': 'u',
 'label': 'c',
 'odor': 'p',
 'population': 's',
 'ring-number': 'o',
 'ring-type': 'p',
 'spore-print-color': 'k',
 'stalk-color-above-ring': 'w',
 'stalk-color-below-ring': 'w',
 'stalk-root': 'e',
 'stalk-shape': 'e',
 'stalk-surface-above-ring': 's',
 'stalk-surface-below-ring': 's',
 'veil-color': 'w',
 'veil-type': 'p'}
```
```python
model.predict_topk(data[0], k=2, output_type='probability')
```
```
Columns:
	id	int
	class	int
	probability	float

Rows: 2

Data:
+----+-------+---------------+
| id | class |  probability  |
+----+-------+---------------+
| 0  |   1   | 0.57558375597 |
| 0  |   0   | 0.42441624403 |
+----+-------+---------------+
[2 rows x 3 columns]
```

In C++, the equivalent code would be (minus the CSV parsing):
```cpp
// Assuming this is in a sibling directory of turicreate,
// named "predict.cpp", run the following:
//
// In turicreate:
//
// ./configure
// cd release/src/unity
// make -j8
//
// In this directory:
// (Note: I had to copy some build flags from turicreate to get it to compile.)
//
// c++ -std=c++11 -stdlib=libc++ -I../turicreate/src -I../turicreate/src/platform -I../turicreate/deps/local/include -I../turicreate/src/external/armadillo/include -DHASH_FOR_UINT128_DEFINED -DHASH_FOR_INT128_DEFINED -L../turicreate/release/src/unity -lunity_shared predict.cpp ../turicreate/release/src/unity/libunity.a
// LD_LIBRARY_PATH=../turicreate/release/src/unity ./a.out

// standard C++ includes
#include <iostream>

// Turi Create includes
#include <logger/assertions.hpp>
#include <logger/logger.hpp>
#include <unity/lib/unity_global.hpp>
#include <unity/server/registration.hpp>
#include <unity/toolkits/supervised_learning/boosted_trees.hpp>

int main(int argc, char **argv) {
  // set up registry of named functions and models
  // used for rehydrating saved models
  auto fn_reg = turi::toolkit_function_registry();
  turi::register_functions(fn_reg);
  auto class_reg = turi::toolkit_class_registry();
  turi::register_models(class_reg);

  // create a global context
  auto global = turi::unity_global(&fn_reg, &class_reg);

  // load the model from disk
  auto loaded_model = global.load_model("mushroom.tcmodel");
  std::string name = turi::variant_get_value<turi::flexible_type>(loaded_model["model_name"]);
  std::cout << "Loaded model: " << name << std::endl;
  auto model = std::dynamic_pointer_cast<turi::supervised::xgboost::boosted_trees_classifier>(
    turi::variant_get_value<std::shared_ptr<turi::model_base>>(loaded_model["model"])
  );

  // set up input
  // (note: can do batch prediction by adding more than one row to rows)
  std::vector<turi::flexible_type> rows;
  turi::flex_dict row;
  row.push_back(std::make_pair("bruises?", "t"));
  row.push_back(std::make_pair("cap-color", "n"));
  row.push_back(std::make_pair("cap-shape", "x"));
  row.push_back(std::make_pair("cap-surface", "s"));
  row.push_back(std::make_pair("gill-attachment", "f"));
  row.push_back(std::make_pair("gill-color", "k"));
  row.push_back(std::make_pair("gill-size", "n"));
  row.push_back(std::make_pair("gill-spacing", "c"));
  row.push_back(std::make_pair("habitat", "u"));
  row.push_back(std::make_pair("odor", "p"));
  row.push_back(std::make_pair("population", "s"));
  row.push_back(std::make_pair("ring-number", "o"));
  row.push_back(std::make_pair("ring-type", "p"));
  row.push_back(std::make_pair("spore-print-color", "k"));
  row.push_back(std::make_pair("stalk-color-above-ring", "w"));
  row.push_back(std::make_pair("stalk-color-below-ring", "w"));
  row.push_back(std::make_pair("stalk-root", "e"));
  row.push_back(std::make_pair("stalk-shape", "e"));
  row.push_back(std::make_pair("stalk-surface-above-ring", "s"));
  row.push_back(std::make_pair("stalk-surface-below-ring", "s"));
  row.push_back(std::make_pair("veil-color", "w"));
  row.push_back(std::make_pair("veil-type", "p"));
  rows.push_back(row);

  // make prediction
  // (returns a gl_sframe)
  turi::gl_sframe result = model->fast_predict_topk(
      // input data - 1 row per prediction
      rows,

      // output type - see https://github.com/apple/turicreate/blob/3490286b27ff5d79cb90d09fe026d5671ce990c7/src/unity/toolkits/supervised_learning/supervised_learning.hpp#L47
      "probability",

      // missing value action - see https://github.com/apple/turicreate/blob/3490286b27ff5d79cb90d09fe026d5671ce990c7/src/ml_data/ml_data_column_modes.hpp#L27
      "error",

      // k (returns top k predictions) -- must be <= the number of classes
      2
  );
  std::string row_str = rows[0].to<std::string>(); // turn first row of input into a string
  std::cout << "Input to model: " << row_str << std::endl;
  std::cout << "Predicted: " << result << std::endl;
}
```
And the output from C++ looks like:
```
Loaded model: boosted_trees_classifier
Input to model: {"bruises?":"t", "cap-color":"n", "cap-shape":"x", "cap-surface":"s", "gill-attachment":"f", "gill-color":"k", "gill-size":"n", "gill-spacing":"c", "habitat":"u", "odor":"p", "population":"s", "ring-number":"o", "ring-type":"p", "spore-print-color":"k", "stalk-color-above-ring":"w", "stalk-color-below-ring":"w", "stalk-root":"e", "stalk-shape":"e", "stalk-surface-above-ring":"s", "stalk-surface-below-ring":"s", "veil-color":"w", "veil-type":"p"}
Predicted: 
Columns:
    id	integer
    class	integer
    probability	float
Rows: 2
Data:
+----------------+----------------+----------------+
| id             | class          | probability    |
+----------------+----------------+----------------+
| 0              | 1              | 0.575584       |
| 0              | 0              | 0.424416       |
+----------------+----------------+----------------+
[2 rows x 3 columns]
```

**Comment by znation** [comment]
I'll keep this issue open for now to track adding the above code sample to the C++ API docs, and restructuring the API docs to present some high level use cases first.

**Comment by CogMachines** [comment]
Thanks for the late night work!  I was able to get it the libs linked in.   I made a custom binaries and header archive so I could deploy the build across various machines.  The code is closed source, but if you are interested in taking a look at the implimentation, I would be glad to share it with you.  It is great to be able to use the same codebase across mobile, desktop and server!

**Comment by srikris** [comment]
Out of curiosity, do you mind telling us which feature in your app was powered by the model?

**Comment by CogMachines** [comment]
The classification of musical symbols and alphanumeric characters is done using the trained model.  I am not using the boosted_trees_classifier that was in the example code, but a different one.  The classified symbol along with some morphological analysis and contextual data is used to the convert the users handwriting into music notation.  Before I integerated CoreML, I was using my own nearest neighbor algorithm using DTW and the distance function.  The upside to the kNN model was being able to allow the user to train the model on iOS.  The dowsides were the usual issues with kNN.  The CoreML model is much more accurate.  I had previously tried out a CNN model using embedded Torch.  I have actually created models that worked in the iOS app using SVM, Vowpal Wabbit and more.  The previous CNN model was really too large to ship.  Being able to use TuriCreate and train with transfer learning is a hug plus.  My deployed model is only 5MB in size.  If you want an in-person demo, I am located in Seattle as well.
